### PR TITLE
Type _sympy/functions.py [1/n]

### PR DESCRIPTION
**Pull request by bobrenjc93**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #136205

Signed-off-by: Bob Ren <bobren@fb.com>

I was chatting with @jamesjwu about strategies to learn the code and he suggested adding types to some files. This stack of PRs adds types to _sympy/functions.py

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

File: torch/utils/_sympy/functions.py [hunk 1]
```diff
@@ -76,26 +79,29 @@
...
-    def inner(*args):
+    def inner(*args: Any) -> T:
         r = f(*args)
         if any(isinstance(a, sympy.Float) for a in args) and not isinstance(
             r, sympy.Float
         ):
-            r = sympy.Float(float(r))
+            try:
+                r = sympy.Float(r)
+            except TypeError:
```
Comment (ezyang) [inline torch/utils/_sympy/functions.py:91]: What's going on here?

**Comment by ezyang** [comment]
Meta-comment: typing code that interacts with Sympy is often quite difficult, as Sympy is not very static type friendly. Something to be aware of; sometimes the juice is not worth the squeeze.

**Review by aorenste — COMMENTED** [review]
I'm not sure if this is meant to completely annotate the types in this file, but if it is you should remove the "mypy: allow-untyped-defs" from the top.

File: torch/utils/_sympy/functions.py [hunk 2]
```diff
@@ -162,28 +172,27 @@ class FloorDiv(sympy.Function):
...
+    def base(self) -> sympy.Basic:
         return self.args[0]
 
     @property
-    def divisor(self):
+    def divisor(self) -> sympy.Basic:
         return self.args[1]
 
-    def _sympystr(self, printer):
+    def _sympystr(self, printer: Any) -> str:
```
Comment (ezyang) [inline torch/utils/_sympy/functions.py:187]: should be sympy.printing.printer.Printer I think

File: torch/utils/_sympy/functions.py [hunk 3]
```diff
@@ -76,26 +79,29 @@
...
+                r = sympy.Float(r)
+            except TypeError:
+                pass
         return r
 
     return inner
 
 
-def fuzzy_eq(x, y):
+def fuzzy_eq(x: Any, y: Any) -> Optional[bool]:
```
Comment (ezyang) [inline torch/utils/_sympy/functions.py:98]: Any here should be Optional[bool]

**Comment by ezyang** [comment]
Blocked on https://github.com/pytorch/pytorch/pull/136205/files#r1763667817

**Comment by bobrenjc93** [comment]
Sorry @ezyang @aorenste I'm still getting used to the GH workflow and didn't mean to publish this PR. My typical workflow is to send draft PRs and only after CI passes, do I publish it. But GH publishes the PR when you add reviewers (which isn't the case in phabricator).

File: torch/utils/_sympy/functions.py (hunk 1 above)
Comment (bobrenjc93) [reply torch/utils/_sympy/functions.py]: fixed

**Comment by aorenste** [comment]
> Sorry @ezyang @aorenste I'm still getting used to the GH workflow and didn't mean to publish this PR. My typical workflow is to send draft PRs and only after CI passes, do I publish it. But GH publishes the PR when you add reviewers (which isn't the case in phabricator).

You can use the ghstack option `--draft` to publish as a draft.

**Comment by bobrenjc93** [comment]
> > Sorry @ezyang @aorenste I'm still getting used to the GH workflow and didn't mean to publish this PR. My typical workflow is to send draft PRs and only after CI passes, do I publish it. But GH publishes the PR when you add reviewers (which isn't the case in phabricator).
> 
> You can use the ghstack option `--draft` to publish as a draft.

That's what I did! But it seems like adding reviewers automatically promotes the PR from draft status to in review.

File: torch/utils/_sympy/functions.py [hunk 4]
```diff
@@ -19,6 +30,8 @@
 from .numbers import int_oo
 
 
+T = TypeVar("T", bound=SupportsFloat)
```
Comment (Skylion007) [inline torch/utils/_sympy/functions.py:33]: Unlikely this TypeVar should be public, should probably begin with `_T` ('in both variable and name)` so it's not exported.

**Review by Skylion007 — APPROVED** [review]
Other than the nit: glad to see the type ignores disappear.

File: torch/utils/_sympy/functions.py [hunk 5]
```diff
@@ -76,9 +89,9 @@
 ]
 
 
-def _keep_float(f):
+def _keep_float(f: Callable[..., T]) -> Callable[..., Union[T, sympy.Float]]:
```
Comment (jamesjwu) [inline torch/utils/_sympy/functions.py:92]: If T supports float already, what does it mean for the return type of the callable to be Union(T, sympy.Float)?

**Review by jamesjwu — APPROVED** [review]
Good to go, but do change T to be private so it's not exported

**Pull request closed without merging**